Testsys status

ecpullen opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**

Closes #2422 
Closes #2423 
Closes #2572 

**Description of changes:**

This pr contains a set of commits that are designed to improve the user experience with `cargo make test`

```
e19242b1 testsys: Provide status output options
58b3eee9 testsys: Improve delete filtering control
a219ea5d testsys: Improve filtering of `watch-test`
41486b7a testsys: Add unique suffix to vm resource
b28f0551 testsys: Add unique suffix to ec2 instances
3ec9cae0 testsys: Add `testsys/test-type` label
caeae492 testsys: Add `BUILD ID` to status output.
43635515 testsys: Add label for build id
```

`e19242b1`
Allows users to customize the status output to fit their screen size (or json).

`58b3eee9`
Changed `cargo make clean-test` to delete all passing tests. Resources will automatically be cleaned up, this reduces the execution time of `cargo make clean-test` significantly. `cargo make clean-test-all` is now used to delete everything from a cluster.

 `a219ea5d`
Changed `cargo make watch-test` to only show test status, `cargo make watch-test-all` can be used to show all tests and resources. Also, `cargo make watch-test-all --passed` can be used to watch all passed tests, `--failed` and `--running` can also be used (separately).

`41486b7a`, `b28f0551`
Instances and VM resources are now given a suffix to their name. This prevents collisions with a resource that is being deleted.

`3ec9cae0`
Adds an additional label for `SelectionParams` filtering, `testsys/test-type`

`caeae492`
Adds a column to `cargo make watch-test` for the `BUILD ID` that is being tested.

`43635515`
Adds a label for the build id that is being tested.

**Testing done:**

Sample output:
`cargo make watch-test`

```
NAME                          TYPE       STATE         PASSED       SKIPPED       FAILED      LAST UPDATE                 BUILD ID
 x86-64-aws-k8s-124-test       Test       waiting                                               2023-01-04T14:32:25Z       3ebb33a9
 ```

`cargo make watch-test-all`

```
NAME                                  TYPE         STATE          PASSED     SKIPPED     FAILED    LAST UPDATE               BUILD ID
 x86-64-aws-k8s-124                    Resource     completed                                        2023-01-04T14:32:28Z     3ebb33a9
 x86-64-aws-k8s-124-instances-kpsq     Resource     completed                                        2023-01-04T14:32:36Z     3ebb33a9
 x86-64-aws-k8s-124-test               Test         inProgress     0          0           0          2023-01-04T14:33:41Z     3ebb33a9
```

`cargo make watch-test-all -o wide`

```
 NAME                               TYPE      STATE       PASSE   SKIPPE   FAILE  STATUS               LAST UPDATE             BUILD ID
 x86-64-aws-k8s-124                 Resourc   completed                            Cluster ready        2023-01-04T14:32:28Z   3ebb33a9
 x86-64-aws-k8s-124-instances-kps   Resourc   completed                            Instance(s) create   2023-01-04T14:32:36Z   3ebb33a9
 x86-64-aws-k8s-124-test            Test      inProgres   0       0        0       Starting Test        2023-01-04T14:36:12Z   3ebb33a9
 ```
 
`cargo make watch-test -o wide`

 ```
  NAME                        TYPE     STATE          PASSED     SKIPPED    FAILED   STATUS           LAST UPDATE              BUILD ID
 x86-64-aws-k8s-124-test     Test     inProgress     0          0          0         Starting Test    2023-01-04T14:36:42Z    3ebb33a9
```

`cargo make watch-test-all -o narrow`
```
 NAME                                         TYPE               STATE                PASSED           SKIPPED           FAILED
 x86-64-aws-k8s-124                           Resource           completed
 x86-64-aws-k8s-124-instances-kpsq            Resource           completed
 x86-64-aws-k8s-124-test                      Test               inProgress           0                0                 0
 ```

`cargo make purge-test`

```
$ cargo make purge-test
[cargo-make] INFO - cargo make 0.36.1
[cargo-make] INFO - Build File: Makefile.toml
[cargo-make] INFO - Task: purge-test
[cargo-make] INFO - Profile: development
[cargo-make] INFO - Running Task: setup
[cargo-make] INFO - Running Task: fetch-sources
[cargo-make] INFO - Running Task: test-tools
[cargo-make] INFO - Running Task: reset-test
[2023-01-06T14:58:07Z INFO  testsys::delete] Deleting all tests and resources
Starting delete for x86-64-aws-k8s-124-test
Delete finished for x86-64-aws-k8s-124-test
Starting delete for x86-64-aws-k8s-124
Delete finished for x86-64-aws-k8s-124
[2023-01-06T14:58:27Z INFO  testsys::delete] Delete finished
[cargo-make] INFO - Running Task: uninstall-test
[2023-01-06T14:58:28Z INFO  testsys::uninstall] testsys components were successfully uninstalled.
[cargo-make] INFO - Build Done in 23.04 seconds.
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

ecpullen committed:
testsys: Add label for build id

ecpullen committed:
testsys: Add `BUILD ID` to status output.

ecpullen committed:
testsys: Add `testsys/test-type` label

ecpullen committed:
testsys: Add unique suffix to ec2 instances

ecpullen committed:
testsys: Add unique suffix to vm resource

ecpullen committed:
testsys: Improve filtering of `watch-test`

ecpullen committed:
testsys: Improve delete filtering control

ecpullen committed:
testsys: Provide status output options

webern commented on Makefile.toml:
```diff
@@ -1423,9 +1434,25 @@ script = [
 ]
 
-# This task will call watch on the `status` testsys command to show the results of a test.
+# This task will call watch on the `status` testsys command to show the results of all tests.
+# To see all passed tests use `cargo make watch-test --passed`
+# To see all failed tests use `cargo make watch-test --failed`
+# To see all incomplete tests use `cargo make watch-test --running`
 [tasks.watch-test]
 dependencies = ["test-tools"]
+script = [
+   '''
+   set -eu
+   export PATH="${BUILDSYS_TOOLS_DIR}/bin:${PATH}"
+   watch -- testsys ${CARGO_MAKE_TESTSYS_ARGS} status --test ${@}
+   '''
+]
```
Should the watch be optional? Rename the task and observe a `--watch` flag?

ecpullen replied:
I don't think so. Having that watch built in is super helpful. Users can do `cargo make testsys status` if they don't want it in `watch`.

etungsten commented on Makefile.toml:
```diff
@@ -1401,9 +1401,20 @@ script = [
 ]
 
-# This task will clear all tests and resources from the testsys cluster.
+# This task will clear all passed tests from the testsys cluster.
 [tasks.clean-test]
 dependencies = ["test-tools"]
+script = [
```
I think we should keep the previous behavior of `clean-test` and introduce a new task `clean-passed-test` instead. This way we keep backwards compatibility and I think it's probably more clear that way as well.

ecpullen replied:
I think we want to keep resource deletion separate (`clean-test-all`) since it adds a significant amount of blocking time for ec2 resources and they will automatically be cleaned up. Should we have a separate task `cargo make clean-test-passed` or just `cargo make clean-test --passed`?

etungsten replied:
I think the new `cargo make clean-test --passed`  makes sense. Thanks!

bcressey commented:
It looks like the alignment is off between the column labels and the individual rows in most of the sample output. Can that be fixed?

bcressey commented on Makefile.toml:
```diff
@@ -1401,14 +1401,28 @@ script = [
+    '''
+]
+
+# This task will clear all tests and resources from the testsys cluster.
+[tasks.clean-test-all]
+dependencies = ["test-tools"]
+script = [
+    '''
+    set -eu
+    export PATH="${BUILDSYS_TOOLS_DIR}/bin:${PATH}"
+    testsys ${CARGO_MAKE_TESTSYS_ARGS} delete ${@}
```
The existing `Makefile.toml` target convention is that `clean-x` targets do a full clean of the "x" resource.

The overall intent is that `cargo make clean` will put the repo into a state that's pretty close to a fresh checkout. The exceptions are things like the cache for Go modules and Rust crates.

It looks like the goal here is to avoid a full `testsys delete` unless a more extreme reset is requested. That's similar to the existing behavior that preserves cached modules - the developer *probably* doesn't want to force those to be downloaded again, and it *probably* won't change the results relative to a full reset.

I'd suggest something like this:
* add `clean-test` to the overall `clean` task, so that `cargo make clean` also cleans up passed and failed tests by default
* rename `clean-test-all` to `purge-test` or `reset-test`
* add an overall `purge` or `reset` target, so that `cargo make purge` does a full clean and then additionally removes the clusters, and perhaps cleans up the Go / Rust caches

ecpullen replied:
I think we will want to wait on adding `clean-test` to `clean` since `clean-test` will fail if testsys is not setup.

ecpullen committed:
testsys: Add `purge-test` task

`cargo make purge-test` will delete all tests, resources, and the
namespace in a testsys cluster.
